packages/eslint-plugin/src/rules/no-unnecessary-type-parameters.ts

```diff
@@ -0,0 +1,346 @@
+import type { TSESTree } from '@typescript-eslint/utils';
+import * as tsutils from 'ts-api-utils';
+import * as ts from 'typescript';
+
+import type { MakeRequired } from '../util';
+import { createRule, getParserServices } from '../util';
+
+type NodeWithTypeParameters = MakeRequired<
+  ts.SignatureDeclaration | ts.ClassLikeDeclaration,
+  'typeParameters'
+>;
+
+export default createRule({
+  defaultOptions: [],
+  meta: {
+    docs: {
+      description: 'Disallow type parameters that only appear once',
+      recommended: 'strict',
+      requiresTypeChecking: true,
+    },
+    messages: {
+      sole: 'Type parameter {{name}} is used only once.',
+    },
+    schema: [],
+    type: 'problem',
+  },
+  name: 'no-unnecessary-type-parameters',
+  create(context) {
+    const parserServices = getParserServices(context);
+    return {
+      [[
+        'ArrowFunctionExpression[typeParameters]',
+        'ClassDeclaration[typeParameters]',
+        'ClassExpression[typeParameters]',
+        'FunctionDeclaration[typeParameters]',
+        'FunctionExpression[typeParameters]',
+        'TSCallSignatureDeclaration[typeParameters]',
+        'TSConstructorType[typeParameters]',
+        'TSDeclareFunction[typeParameters]',
+        'TSEmptyBodyFunctionExpression[typeParameters]',
+        'TSFunctionType[typeParameters]',
+        'TSMethodSignature[typeParameters]',
+      ].join(', ')](node: TSESTree.FunctionLike): void {
+        const tsNode = parserServices.esTreeNodeToTSNodeMap.get(
+          node,
+        ) as NodeWithTypeParameters;
+
+        const checker = parserServices.program.getTypeChecker();
+        const counts = countTypeParameterUsage(checker, tsNode);
+
+        for (const typeParameter of tsNode.typeParameters) {
+          const identifierCounts = counts.get(typeParameter.name);
+          if (
+            !identifierCounts ||
+            identifierCounts.general > 2 ||
+            identifierCounts.return > 2 ||
+            (identifierCounts.general === 2 && identifierCounts.return) ||
+            (identifierCounts.return === 2 && identifierCounts.general)
+          ) {
+            continue;
+          }
+
+          context.report({
+            data: {
+              name: typeParameter.name.text,
+            },
+            node: parserServices.tsNodeToESTreeNodeMap.get(typeParameter),
+            messageId: 'sole',
+          });
+        }
+      },
+    };
+  },
+});
+
+interface IdentifierCounts {
+  general: number;
+  return: number;
+}
+
+type IdentifierArea = keyof IdentifierCounts;
+
+/**
+ * Count uses of type parameters in inferred return types.
+ * We need to resolve and analyze the inferred return type of a function
+ * to see whether it contains additional references to the type parameters.
+ * For classes, we need to do this for all their methods.
+ */
+function countTypeParameterUsage(
```

Comment:
What do you see them as covering that's not covered by the rule?

We generally lean away from function unit tests in this repo because of how comprehensive the rule tests tend to be.

Reply:
The rule only tests whether the count is >= 2. A unit test could check whether the count is actually correct. I could easily imagine off-by-ones in this function that we don't catch because they happen to be 2 vs. 3.

Reply:
That makes sense. I'm going to opt out though - this is such a large percentage of the rule, any unit test for this specific function is going to look roughly the same as the rule test. And if there's any case such as an off-by-one that we need to test, it's just about as straightforward to rule test it as it would be to unit test. 

Long term, if we end up having some kind of "type manager" equivalent to the [scope manager](https://typescript-eslint.io/packages/scope-manager), I could see this kind of logic getting extracted into an individually tested package.

Thanks though!